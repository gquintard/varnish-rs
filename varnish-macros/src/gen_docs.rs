//! Code to generate automated documentation into a file

use std::fmt::Write;
use std::fs;
use std::path::Path;

use crate::model::{FuncInfo, FuncType, ParamInfo, ParamKind, ParamType, ParamTypeInfo, VmodInfo};

// Small helpers to write to a string without checking the result

macro_rules! wrt {
    ($docs:expr, $($arg:tt)*) => {
        let _ = write!($docs, $($arg)*);
    };
}
macro_rules! ln {
    ($docs:expr, $($arg:tt)*) => {
        let _ = writeln!($docs, $($arg)*);
    };
}

/// Generate documentation for the VMOD and save it to a file
pub fn generate_docs(info: &VmodInfo) {
    let Some(ref doc_file) = info.params.docs else {
        return; // doc file is not set, skipping
    };
    let Ok(dir) = std::env::var("CARGO_MANIFEST_DIR") else {
        panic!("Unable to get the CARGO_MANIFEST_DIR env var to save documentation, you may need to remove the `docs` parameter from the `#[vmod]` attribute");
    };
    let doc_file = Path::new(&dir).join(doc_file);
    let docs = gen_doc_content(info);
    if let Err(e) = fs::write(doc_file.as_path(), docs) {
        panic!(
            "Unable to save documentation to file {}: {e}",
            doc_file.display()
        );
    }
}

/// Generate documentation for the VMOD as a single string
pub fn gen_doc_content(info: &VmodInfo) -> String {
    let mut docs = String::new();

    // Initial warning to not edit this file
    ln!(
        docs,
        r#"<!--

   !!!!!!  WARNING: DO NOT EDIT THIS FILE!

   This file was generated from the Varnish VMOD source code.
   It will be automatically updated on each build.

-->"#
    );

    // generate documentation
    ln!(docs, "# Varnish Module (VMOD) `{}`", info.ident);
    write_docs(&mut docs, &info.docs, "#");
    ln!(
        docs,
        r#"
```vcl
// Place import statement at the top of your VCL file
// This loads vmod from a standard location
import {ident};

// Or load vmod from a specific file
import {ident} from "path/to/lib{ident}.so";
```"#,
        ident = info.ident
    );

    for func in &info.funcs {
        if !matches!(func.func_type, FuncType::Function) {
            continue;
        }
        write_function(&mut docs, "###", "Function", func);
    }

    for obj in &info.objects {
        ln!(docs, "\n### Object `{}`", obj.ident);

        write_docs(&mut docs, &obj.docs, "###");
        write_function(&mut docs, "####", &obj.ident, &obj.constructor);
        for method in &obj.funcs {
            write_function(&mut docs, "####", "Method", method);
        }
    }

    docs
}

fn write_function(mut docs: &mut String, prefix: &str, obj_or_typ: &str, func: &FuncInfo) {
    let user_args = get_user_args(func);

    if matches!(func.func_type, FuncType::Constructor) {
        ln!(
            docs,
            r#"
```vcl
// Create a new instance of the object in your VCL init function
sub vcl_init {{
    new {ident} = {obj_or_typ}.{sig};
}}
```"#,
            ident = func.ident,
            sig = fn_sig(func, &user_args),
        );
    } else {
        wrt!(docs, "\n{prefix} {obj_or_typ} ");
        ln!(&mut docs, "{}", fn_sig(func, &user_args));
    }

    write_docs(docs, &func.docs, prefix);

    // List of arguments are printed only if any of them have documentation
    if user_args.iter().any(|(arg, _)| !arg.docs.is_empty()) {
        ln!(docs, "");
        for (arg, ty) in &user_args {
            wrt!(docs, "* `{}`:", bracketed_name(arg, ty));
            if arg.docs.is_empty() {
                ln!(docs, "");
            } else {
                write_docs(docs, &arg.docs, prefix);
            }
        }
    }
}

fn fn_sig(func: &FuncInfo, user_args: &Vec<(&ParamTypeInfo, &ParamInfo)>) -> String {
    let mut res = String::new();
    let is_md_txt = matches!(
        func.func_type,
        FuncType::Function | FuncType::Method | FuncType::Event
    );
    if is_md_txt {
        wrt!(res, "`");
    }
    if matches!(func.func_type, FuncType::Function | FuncType::Method) {
        wrt!(res, "{} ", func.returns.value_type().to_vcc_type());
    }
    wrt!(res, "{}(", func.ident);
    let mut first = true;
    for (arg, ty) in user_args {
        if first {
            first = false;
        } else {
            wrt!(res, ", ");
        }
        res.push_str(&bracketed_name(arg, ty));
        if !ty.default.is_null() {
            wrt!(res, " = {}", ty.default);
        }
    }
    wrt!(res, ")");
    if is_md_txt {
        wrt!(res, "`");
    }
    res
}

fn get_user_args(func: &FuncInfo) -> Vec<(&ParamTypeInfo, &ParamInfo)> {
    func.args
        .iter()
        .filter_map(|v| {
            if let ParamType::Value(ref val) = v.ty {
                Some((v, val))
            } else {
                None
            }
        })
        .collect()
}

fn bracketed_name(arg: &ParamTypeInfo, ty: &ParamInfo) -> String {
    let vcc = ty.ty_info.to_vcc_type();
    let ident = &arg.ident;
    if matches!(ty.kind, ParamKind::Optional) {
        format!("[{vcc} {ident}]")
    } else {
        format!("{vcc} {ident}")
    }
}

/// Print user provided documentation, indenting the `## ...` subsections with a prefix
fn write_docs(docs: &mut String, user_doc: &str, prefix: &str) {
    if !user_doc.is_empty() {
        if user_doc.contains('#') {
            // If the user provided documentation contains a `#`, we need to indent the subsections
            let mut user_doc = user_doc.replace("\n#", &format!("\n#{prefix}"));
            if user_doc.starts_with('#') {
                user_doc.insert_str(0, prefix);
            }
            ln!(docs, "\n{user_doc}");
        } else {
            ln!(docs, "\n{user_doc}");
        }
    }
}
