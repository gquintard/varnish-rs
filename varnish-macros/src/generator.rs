//! The main generator for the varnish vmod.

use std::ffi::CString;
use std::fmt::Write as _;

use proc_macro2::TokenStream;
use quote::quote;
use serde_json::{json, Value};
use sha2::{Digest as _, Sha256};
use syn::{Item, ItemMod, Path};

use crate::gen_func::FuncProcessor;
use crate::gen_objects::ObjProcessor;
use crate::model::VmodInfo;
use crate::names::{ForceCstr, Names, ToIdent};

pub fn render_model(mut item_mod: ItemMod, info: &VmodInfo) -> TokenStream {
    let output = Generator::render(info);
    item_mod
        .content
        .as_mut()
        .unwrap()
        .1
        .insert(0, Item::Verbatim(output));

    quote! { #item_mod }
}

#[derive(Debug, Default)]
pub struct Generator {
    pub names: Names,
    pub file_id: CString,
    pub functions: Vec<FuncProcessor>,
    pub objects: Vec<ObjProcessor>,
}

/// See also <https://varnish-cache.org/docs/7.5/reference/vmod.html>
impl Generator {
    pub fn render(vmod: &VmodInfo) -> TokenStream {
        let mut obj = Self {
            names: Names::new(&vmod.ident),
            file_id: Self::calc_file_id(vmod).force_cstr(),
            ..Self::default()
        };
        for info in &vmod.funcs {
            obj.functions.push(FuncProcessor::from_info(
                obj.names.to_func(info.func_type, &info.ident),
                info,
            ));
        }
        for info in &vmod.objects {
            obj.objects
                .push(ObjProcessor::from_info(obj.names.to_obj(&info.ident), info));
        }
        obj.render_generated_mod(vmod)
    }

    /// Use the entire data model parsed from sources to generate a hash.
    /// Should be somewhat consistent and unique for each set of functions.
    fn calc_file_id(info: &VmodInfo) -> String {
        Sha256::digest(format!("{info:?}").as_bytes())
            .into_iter()
            .fold(String::new(), |mut output, b| {
                let _ = write!(output, "{b:02x}");
                output
            })
    }

    fn gen_priv_structs(tokens: &mut Vec<TokenStream>, name: &str, shared_type: Option<&String>) {
        if let Some(type_name) = shared_type {
            let ident = name.to_ident();
            // The type name is stored as a string, but we already validated we can parse it during the `parse` phase.
            let ty_ident = syn::parse_str::<Path>(type_name).expect("Unable to parse second time");
            let ty_name = type_name.force_cstr();
            // Static methods to clean up the `vmod_priv` object's `T`
            tokens.push(quote! {
               static #ident: vmod_priv_methods = vmod_priv_methods {
                   magic: VMOD_PRIV_METHODS_MAGIC,
                   type_: #ty_name.as_ptr(),
                   fini: Some(vmod_priv::on_fini::<#ty_ident>),
               };
            });
        }
    }

    fn iter_all_funcs(&self) -> impl Iterator<Item = &FuncProcessor> {
        self.functions
            .iter()
            .chain(self.objects.iter().flat_map(|o| o.funcs.iter()))
    }

    fn gen_json(&self) -> String {
        let mod_name = &self.names.mod_name();
        let mut json = Vec::<Value>::new();

        json.push(json! {[
            "$VMOD",
            "1.0",
            mod_name,
            self.names.func_struct_name(),
            self.file_id.to_str().unwrap(),
            // Ohh the irony, this string from VMOD_ABI_Version is the reason
            // why `varnish-sys` must exist. Without it, we could run bindgen
            // from the `varnish` crate directly.  Ohh well.
            //
            // FIXME: figure out a way to assert that the version string used by varnish_macro is the same
            //        as the value accessible by generated code from varnish::ffi::VMOD_ABI_Version.
            //        Currently it seems not possible to do a constant assert at compile time on b-str/c-str equality.
            varnish_sys::ffi::VMOD_ABI_Version.to_str().unwrap(),
            "0",
            "0"
        ]});

        let mut cproto = String::new();
        for obj in &self.objects {
            cproto.push_str(&obj.cproto_typedef_decl);
        }
        for func in self.iter_all_funcs() {
            cproto.push_str(&func.cproto_typedef_decl);
        }
        let _ = write!(cproto, "\nstruct {} {{\n", self.names.func_struct_name());
        for func in self.iter_all_funcs() {
            cproto.push_str(&func.cproto_typedef_init);
        }
        let _ = write!(
            cproto,
            "}};\n\nstatic struct {struct_name} {struct_name};",
            struct_name = self.names.func_struct_name()
        );
        json.push(json! {[ "$CPROTO", cproto ]});

        for func in &self.functions {
            json.push(func.json.clone());
        }

        for obj in &self.objects {
            json.push(obj.json.clone());
        }

        let json = serde_json::to_string_pretty(&json! {json}).unwrap();
        format!("VMOD_JSON_SPEC\u{2}\n{json}\n\u{3}")
    }

    fn render_generated_mod(&self, vmod: &VmodInfo) -> TokenStream {
        let vmod_name_data = self.names.data_struct_name().to_ident();
        let c_name = self.names.mod_name().force_cstr();
        let c_func_name = self.names.func_struct_name().force_cstr();
        let file_id = &self.file_id;
        let mut priv_structs = Vec::new();
        Self::gen_priv_structs(
            &mut priv_structs,
            "PRIV_TASK_METHODS",
            vmod.shared_types.shared_per_task_ty.as_ref(),
        );
        Self::gen_priv_structs(
            &mut priv_structs,
            "PRIV_VCL_METHODS",
            vmod.shared_types.shared_per_vcl_ty.as_ref(),
        );
        let functions = self.iter_all_funcs().map(|f| &f.wrapper_function_body);
        let json = &self.gen_json().force_cstr();
        let export_decls: Vec<_> = self.iter_all_funcs().map(|f| &f.export_decl).collect();
        let export_inits: Vec<_> = self.iter_all_funcs().map(|f| &f.export_init).collect();

        quote!(
            #[allow(
                non_snake_case,
                unused_imports,
                unused_qualifications,
                unused_variables,
            )]
            mod varnish_generated {
                use std::borrow::Cow;
                use std::ffi::{c_char, c_int, c_uint, c_void, CStr};
                use std::ptr::null;
                use varnish::ffi::{
                    VCL_BACKEND,
                    VCL_BOOL,
                    VCL_DURATION,
                    VCL_INT,
                    VCL_IP,
                    VCL_PROBE,
                    VCL_REAL,
                    VCL_STRING,
                    VCL_VOID,
                    VMOD_ABI_Version,
                    VMOD_PRIV_METHODS_MAGIC,
                    vcl_event_e,
                    vmod_priv,
                    vmod_priv_methods,
                    vrt_ctx,
                };
                use varnish::vcl::{Ctx, IntoVCL};
                use super::*;

                #( #priv_structs )*
                #( #functions )*

                #[repr(C)]
                pub struct VmodExports {
                    #(#export_decls,)*
                }

                // #[no_mangle]  // FIXME: no_mangle does not seem to be needed
                pub static VMOD_EXPORTS: VmodExports = VmodExports {
                    #(#export_inits,)*
                };

                #[repr(C)]
                pub struct VmodData {
                    vrt_major: c_uint,
                    vrt_minor: c_uint,
                    file_id: *const c_char,
                    name: *const c_char,
                    func_name: *const c_char,
                    func: *const c_void,
                    func_len: c_int,
                    proto: *const c_char,
                    json: *const c_char,
                    abi: *const c_char,
                }

                unsafe impl Sync for VmodData {}

                // This name must be in the format `Vmod_{modulename}_Data`.
                #[allow(non_upper_case_globals)]
                #[no_mangle]
                pub static #vmod_name_data: VmodData = VmodData {
                    vrt_major: 0,
                    vrt_minor: 0,
                    file_id: #file_id.as_ptr(),
                    name: #c_name.as_ptr(),
                    func_name: #c_func_name.as_ptr(),
                    func_len: size_of::<VmodExports>() as c_int,
                    func: &VMOD_EXPORTS as *const _ as *const c_void,
                    abi: VMOD_ABI_Version.as_ptr(),
                    json: JSON.as_ptr(),
                    proto: null(),
                };

                const JSON: &CStr = #json;
            }
        )
    }
}
