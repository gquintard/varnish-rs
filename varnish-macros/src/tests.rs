use std::fs::read_to_string;
use std::path::PathBuf;

use insta::{assert_snapshot, with_settings};
use proc_macro2::TokenStream;
use quote::quote;
use regex::Regex;
use syn::ItemMod;

use crate::gen_docs::gen_doc_content;
use crate::generator::render_model;
use crate::parser::tokens_to_model;
use crate::parser_utils::remove_attr;

/// Read content of the ../tests/pass directory that should also pass full compilation tests,
/// parse them and create snapshots of the model and the generated code.
#[test]
fn parse_passing_tests() {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/pass/*.rs");
    for file in glob::glob(path.to_str().unwrap()).unwrap() {
        let filepath = file.unwrap();
        let file = filepath.to_str().unwrap();
        eprintln!("Processing file: {file}");
        let src = read_to_string(&filepath).expect("unable to read file");
        let syntax = syn::parse_file(&src).expect("unable to parse file");
        let mut has_vmod = false;
        for item in syntax.items {
            if let syn::Item::Mod(mut item) = item {
                assert!(!has_vmod, "Multiple vmod modules found in file {file}");
                has_vmod = true;
                // FIXME: use this attribute as an arg for the test
                let _arg = remove_attr(&mut item.attrs, "vmod").unwrap();
                let name = format!(
                    "{}_{}",
                    filepath.file_stem().unwrap().to_string_lossy(),
                    item.ident
                );
                test(&name, quote! {}, item);
                // FIXME: pass proper attribute info
                // test(&name, quote! { #arg }, item);
            }
        }
        assert!(has_vmod, "No vmod modules found in file {file}");
    }
}

fn test(name: &str, args: TokenStream, mut item_mod: ItemMod) {
    // panic!("{name} {args:?} {item_mod:?}");
    with_settings!({ snapshot_path => "../../varnish/snapshots", omit_expression => true, prepend_module_to_snapshot => false }, {
        let Ok(info) = tokens_to_model(args, &mut item_mod).map_err(|err| {
            // On error, save the error output as a snapshot and return early.
            let err = err.into_compile_error();
            with_settings!({ snapshot_suffix => "error" }, { assert_snapshot!(name, err) });
        }) else { return };

        with_settings!({ snapshot_suffix => "model" }, { assert_snapshot!(name, format!("{info:#?}")) });
        with_settings!({ snapshot_suffix => "docs" }, { assert_snapshot!(name, gen_doc_content(&info)) });

        let file = render_model(item_mod, &info).to_string();
        let parsed = match syn::parse_file(&file) {
            Ok(v) => v,
            Err(e) => {
                // We still save the generated code, but fail the test. Allows easier error debugging.
                with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, file) });
                panic!("Failed to parse generated code in test {name}: {e}");
            }
        };

        let generated = prettyplease::unparse(&parsed);

        // Use regex to remove "Varnish 7.5.0 eef25264e5ca5f96a77129308edb83ccf84cb1b1" and similar.
        // Also removes any pre-builds and other versions because we assume a double-quote at the end.
        // TODO: Once MSRV is higher than 1.80, use static LazyLock<Regex>
        let re_varnish_ver = Regex::new(r"Varnish \d+\.[-+. 0-9a-z]+").unwrap();
        let code = re_varnish_ver.replace_all(&generated, "Varnish (version) (hash)");
        with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, code) });

        // Extract JSON string
        let pat = "const JSON: &CStr = c\"";
        let json = if let Some(pos) = code.find(pat) {
            let json = &code[pos + pat.len()..];
            json.split("\";\n").next()
        } else {
            None
        };

        let json = &json
            .unwrap_or("")
            .replace("\\\"", "\"")
            .replace("\\u{2}", "\u{2}")
            .replace("\\u{3}", "\u{3}")
            .replace("\\\\", "\\")
            // this is a bit of a hack because the double-escaping gets somewhat incorrectly parsed
            .replace("\\n", "\n");

        with_settings!({ snapshot_suffix => "json" }, { assert_snapshot!(name, json) });
    });
}
