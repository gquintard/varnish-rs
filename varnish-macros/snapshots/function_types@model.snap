---
source: varnish-macros/src/tests.rs
snapshot_kind: text
---
VmodInfo {
    params: VmodParams {
        docs: None,
    },
    ident: "types",
    docs: "",
    funcs: [
        FuncInfo {
            func_type: Function,
            ident: "to_void",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_void_err",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                Default,
                VclError,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_str_err",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                Default,
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_bool",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: Bool,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_bool_dflt",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Number(1),
                            ty_info: Bool,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_bool",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: Bool,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_bool",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    Bool,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_bool",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    Bool,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_duration",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: Duration,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_duration",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: Duration,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_duration",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    Duration,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_duration",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    Duration,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_f64",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: F64,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_f64_dflt",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Number(42.3),
                            ty_info: F64,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_f64",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: F64,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_f64",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    F64,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_f64",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    F64,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_i64",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: I64,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_i64_dflt",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Number(10),
                            ty_info: I64,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_i64",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: I64,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_i64",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    I64,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_i64",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    I64,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_str",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: Str,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_str",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: Str,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_str_dflt",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: String("baz"),
                            ty_info: Str,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_str_dflt",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: String("baz"),
                            ty_info: Str,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_str",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_str",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    Str,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_string",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                String,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_opt_string",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                String,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_string",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                String,
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_opt_string",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                String,
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_probe",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: Probe,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_probe_req",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: Probe,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_probe",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    Probe,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_probe",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    Probe,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_cow_probe",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: ProbeCow,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_cow_probe_req",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: ProbeCow,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_cow_probe",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    ProbeCow,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_cow_probe",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    ProbeCow,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_ip",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: SocketAddr,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "type_ip_req",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "_v",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: SocketAddr,
                        },
                    ),
                },
            ],
            returns: Value(
                Default,
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_ip",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                ParamType(
                    SocketAddr,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_ip",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                ParamType(
                    SocketAddr,
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_vcl_string",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Value(
                VclType(
                    "VCL_STRING",
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "to_res_vcl_string",
            docs: "",
            has_optional_args: false,
            args: [],
            returns: Result(
                VclType(
                    "VCL_STRING",
                ),
                ParamType(
                    Str,
                ),
            ),
        },
        FuncInfo {
            func_type: Function,
            ident: "opt_i64_opt_i64",
            docs: "",
            has_optional_args: true,
            args: [
                ParamTypeInfo {
                    ident: "a1",
                    docs: "",
                    idx: 0,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: I64,
                        },
                    ),
                },
                ParamTypeInfo {
                    ident: "a2",
                    docs: "",
                    idx: 1,
                    ty: Value(
                        ParamInfo {
                            is_optional: true,
                            default: Null,
                            ty_info: I64,
                        },
                    ),
                },
                ParamTypeInfo {
                    ident: "a3",
                    docs: "",
                    idx: 2,
                    ty: Value(
                        ParamInfo {
                            is_optional: false,
                            default: Null,
                            ty_info: I64,
                        },
                    ),
                },
            ],
            returns: Value(
                String,
            ),
        },
    ],
    objects: [],
    shared_types: SharedTypes {
        shared_per_task_ty: None,
        shared_per_vcl_ty: None,
    },
}
